### PR TITLE
fix: prevent to override factor enrollment

### PR DESCRIPTION
**Pull request by leleueri**
fixes AM-1194

### CODE REVIEW

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                resp -> {
+                    String location = resp.headers().get("location");
+                    assertNotNull(location);
+                    assertTrue(location.contains("/error"));
+                    assertTrue(location.contains("factor+already+enrolled"));
+                },
+                HttpStatusCode.FOUND_302, "Found", null);
+    }
+
+  private static User createUser() {
```
Comment (mstasgravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java:335]: createUser() and createUserFactor() are unused

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class MFAEnrollEndpointTest extends RxWebTestBase {
+
```
Comment (mstasgravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java:60]: please fix code formatting

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java [hunk 3]
```diff
@@ -202,6 +220,14 @@ private void saveEnrollment(RoutingContext routingContext) {
         }
     }
 
+    private boolean userAlreadyHasFactor(User endUser, Client client) {
+        final var recoveryCodeFactors = getRecoveryFactorIds(client);
+        return !CollectionUtils.isEmpty(endUser.getFactors()) &&
```
Comment (mstasgravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java:225]: thought: we use CollectionUtils.isNotEmpty(endUser.getFactors())

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java [hunk 4]
```diff
@@ -42,10 +45,12 @@
 import io.vertx.reactivex.ext.web.common.template.TemplateEngine;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.util.CollectionUtils;
 
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
```
Comment (mstasgravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java:53]: 161 line: 
final boolean acceptEnrollment = Boolean.parseBoolean(params.get(ConstantKeys.USER_MFA_ENROLLMENT));

Comment (mstasgravitee) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java:53]: thought: Can we add an error for default in the switch 239 line?

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java (hunk 1 above)
Comment (leleueri) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java:335]: Don't know why it is not used in this version, My branch is up to date and these methods are both used... :thinking:

Comment (mstasgravitee) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java:335]: image
createUser() is unused and createUserFactor() is used only in createUser()

**Review by mstasgravitee — APPROVED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java (hunk 1 above)
Comment (leleueri) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpointTest.java:335]: indeed, was looking to the MFAChallengeEndpointTest class :face_exhaling:

**Pull request merged**